### ISSUE TITLE
Endless re-compile in `spin watch` for Spin Apps that consist of multiple components

### ISSUE THREAD

**Issue by ThorstenHans** [fermyon/spin#2637]
I stumbled upon a bug when using `spin watch` with an Spin App consisting of multiple components of which one is a `static-fileserver`.

## TL;DR;

The `spin watch` command keeps on re-compiles the app in an endless loop because the `static-fileserver` command does not contain a build configuration. According to the implementation of `BuildFilterFactory`, this leads to watching all files for changes. Suggested solutions are mentioned in the corresponding section below.

## Spin App to reproduce this behavior

You can find an Spin App for reproducing the behavior at [github.com/ThorstenHans/tmp--spin-watch-repro](https://github.com/ThorstenHans/tmp--spin-watch-repro). 

### App layout
The app consists of two components a `backend` component which has been created using the `http-rust` template and an `ui` component which has been created using the `static-fileserver` template.

## Actual behavior

When executing `spin watch`, the Spin App will be compiled in an endless loop. 

Logs show that files in `./target/` are causing this behavior:

```console
2024-07-12T10:18:57.011356Z DEBUG spin_cli::commands::watch: spin watch detected changes requiring reload: /Users/thorsten/dev/demos/spin-watch-repro/backend/target/wasm32-wasi/release/backend.wasm
```

## Revisiting the implementation of `spin watch`

I looked into the implementation of `spin watch` and recognized that `create_source_globs` as part of `src/commands/watch/filters.rs` returns an empty vector for the `ui` component: 

[https://github.com/fermyon/spin/blob/a7ffa3ae387d9cb7e33c8409719d80750f5c0751/src/commands/watch/filters.rs#L113-L115](https://github.com/fermyon/spin/blob/a7ffa3ae387d9cb7e33c8409719d80750f5c0751/src/commands/watch/filters.rs#L113-L115)

This leads to the `buildifier` recursively watching all files.   
  

## Suggested solution 1: Updating the `static-fileserver` component template

On way of addressing this behavior could be updating the component template of `static-fileserver` to add the following build configuration (assuming that the source folder for the static fileserver is set to `assets/`:

```Toml
[component.ui]
source = { url = "https://github.com/fermyon/spin-fileserver/releases/download/v0.2.1/spin_static_fs.wasm", digest = "sha256:5f05b15f0f7cd353d390bc5ebffec7fe25c6a6d7a05b9366c86dcb1a346e9f0f" }
files = [{ source = "assets", destination = "/" }]

[component.ui.build]
command = ""
watch = ["assets/**/*"]
```

## Suggested solution 2: Updating the implementation of `spin watch`

Alternatively, the implementation of `spin watch` should be modified to terminate `spin watch` if the Spin App consists of multiple components and one or more components lack either the entire build configuration or specify `watch = []` .

**Comment by itowlson** [comment]
If a component doesn't have a `build.watch` then we should watch nothing.  That's the intent of the `BuildFilterFactory` returning the empty vector.  I guess `watchexec` is taking "empty list" to mean "everything" which is unfortunate design!  (Normally the list is additive.)  I'll have an investigate - thanks for flagging it!

We should not `build.watch` the assets.  All `files` entries are already watched for reload (and trigger only a reload, not a rebuild).  As a temporary workaround, I'd suggest `build.watch`-ing a never-matches pattern because there is never any need to trigger a rebuild for the fileserver component.
